pyomo/contrib/matching/interface.py

```diff
@@ -0,0 +1,79 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and 
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain 
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.common.collections import ComponentMap
+from pyomo.contrib.matching.maximum_matching import maximum_matching
+from pyomo.contrib.matching.block_triangularize import block_triangularize
+from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
+
+class IncidenceGraphInterface(object):
+    """
+    The purpose of this class is to allow the user to easily
+    analyze graphs of variables and contraints in a Pyomo
+    model without constructing multiple PyomoNLPs.
+    """
+
+    def __init__(self, model):
+        self.nlp = PyomoNLP(model)
```

Comment:
I don't really see the value in caching a local PyomoNLP.  Wouldn't it make more sense to just work with whatever the user give us?

Reply:
For this interface class, I think I prefer just working with a Pyomo model. My reasoning was that if the user already has a PyomoNLP that they want to work with, they can call `extract_submatrix_jacobian` themself and use the `maximum_matching` and `block_triangularize` functions directly on the matrix. I cache the PyomoNLP so we can call the methods repeatedly with different components without repeating the work of writing/parsing the nl file. The desire to cache the PyomoNLP then block triangularize many subsets of components was the motivation for creating this class in the first place.

I can think of a couple improvements:
1. Don't cache the PyomoNLP, then at the time the submatrix is needed, construct a smaller PyomoNLP based on only the variables and constraints the user asked for (by attaching these components to a dummy block as references).
2. Don't create a PyomoNLP at all, then construct the sparse matrix with `identify_variables`, or if we need numeric derivatives, Pyomo's AD/symbolic differentiation functions.

For both of these options, we don't need to cache anything, and also don't need access to the user's model, so I don't think they need to be part of this (or any) class at all. We could add additional functions for these in the `block_triangularize` and `maximum_matching` modules. We should at least do this for an implementation that only uses a structural Jacobian.